### ISSUE TITLE
Call Webhook when ratelimit error

### ISSUE THREAD

**Issue by benjy3gg** [bbernhard/signal-cli-rest-api#482]
### Feature Request
Hi, we are using this library in a production environment for sending messages from multiple numbers to customers.
We sometimes have the problem, that we get rate limited.

Is there a way to implement it, so that the signal-cli-rest-api is calling a webhook to notify us that there is a rate limit challenge waiting or any other way to get notified of that?

**Comment by benjy3gg** [comment]
i have the same problem again now.
Ran into some kind of ratelimiting and when i try to send a message to a number i only get the following response:

i am sending a normal v2/send request

`{"message": "Test via Signal API!", "number": "+xxxxx", "recipients": [ "+xxxxxx" ]}`

`{
	"error": "Failed to send message due to rate limiting"
}` -> status code 400 

(was expecting a 429, and the challenge token)

**Comment by bbernhard** [comment]
Which mode are you using? json-rpc?

That error message actually comes from [signal-cli](https://github.com/AsamK/signal-cli). I am wondering, if it has to do with the mode and if the error message somehow is different in those modes.

How many messages do you send that you end up with that rate limiting error? I am trying to reproduce that on my side, but no luck so far.

**Comment by benjy3gg** [comment]
i am using your pysignalclirestapi to send the messages, maybe thats where some of the error data is lost?

**Comment by benjy3gg** [comment]
Tried again with calling the api directly and i also only get the "error": "Failed to send message due to rate limiting", so its not pysignalclirestapi omitting some error details.

**Comment by bbernhard** [comment]
The `Failed to send message due to rate limiting` error definitely comes from signal-cli directly. Unfortunately I am not able to provoke that situation (tried to send messages in an endless loop, but for some reason I do not run into that issue).

In case you can reproduce that issue, I can add some debug logging, so that we can (hopefully) narrow the issue down.

**Comment by benjy3gg** [comment]
yes, i can reproduce the issue. would be perfect if you could implement the logging!

**Comment by bbernhard** [comment]
Here is the image with debug logging: `bbernhard/signal-cli-rest-api:0.146-dev`

In order to enable debug logging, please call:
`curl -X POST -H "Content-Type: application/json" -d '{"logging": {"level": "debug"}}' 'http://127.0.0.1:8080/v1/configuration'`

As soon as you've enabled debug logging, you should see additional log output in the docker log file, when you perform a REST API call.

Please be aware that the debug logging setting isn't persisted. That means, if you (or docker-compose) restarts the docker container, you need to use the above REST API call again to enable debug logging.

**Comment by benjy3gg** [comment]
tested it, but i dont see any more debug logging than before:
<img width="806" alt="image" src="https://github.com/bbernhard/signal-cli-rest-api/assets/458675/2ce4d300-cc0e-4372-9994-9cd9fb2d72e0">

**Comment by benjy3gg** [comment]
the build seems to be stuck, can you retrigger it?

**Comment by bbernhard** [comment]
> the build seems to be stuck, can you retrigger it?

:+1: 

When it is finished, `bbernhard/signal-cli-rest-api:0.148-dev` should be up on dockerhub.

**Comment by bbernhard** [comment]
Yeah, it looks like this is all we get back from signal-cli when the message sending fails. :/

Just to be sure, I've extended the debug logging a bit more and now also log all the messages that we get from signal-cli asynchronously. _Maybe_ the challenge token is communicated that way.

Could you please try `bbernhard/signal-cli-rest-api:0.149-dev` and enable the debug logging again.

**Comment by benjy3gg** [comment]
This is what i got back.
`{
    "jsonrpc": "2.0",
    "error": {
        "code": -5,
        "message": "Failed to send message due to rate limiting",
        "data": {
            "response": {
                "timestamp": 1711045916695,
                "results": [
                    {
                        "recipientAddress": {
                            "uuid": "xxxxx-xxxx-xxxx-xxxx-xxxxxxxxx",
                            "number": "+xxxxxxx"
                        },
                        "type": "RATE_LIMIT_FAILURE",
                        "token": "xxxxxxxxx-xxxx-4b5f-xxxx-xxxxxxxx",
                        "retryAfterSeconds": 86400
                    }
                ]
            }
        }
    },
    "id": "xxxxxxx-73ef-xxxxxxx-xxxx-xxxxxxxxxxx"
}`

using the token, i was able to complete the challenge!
Could you add this somehow to the normal error message?

**Comment by bbernhard** [comment]
> This is what i got back. `{ "jsonrpc": "2.0", "error": { "code": -5, "message": "Failed to send message due to rate limiting", "data": { "response": { "timestamp": 1711045916695, "results": [ { "recipientAddress": { "uuid": "xxxxx-xxxx-xxxx-xxxx-xxxxxxxxx", "number": "+xxxxxxx" }, "type": "RATE_LIMIT_FAILURE", "token": "xxxxxxxxx-xxxx-4b5f-xxxx-xxxxxxxx", "retryAfterSeconds": 86400 } ] } } }, "id": "xxxxxxx-73ef-xxxxxxx-xxxx-xxxxxxxxxxx" }`
> 
> using the token, i was able to complete the challenge! Could you add this somehow to the normal error message for the api call

Could you please add the complete log line (including the `level = debug msg=...`). I've now added so many debug lines that I do not know where that payload is actually returned.

**Comment by benjy3gg** [comment]
it was in this one:
`level=debug msg="json-rpc received data: `

**Comment by bbernhard** [comment]
Thanks!

So it looks like that the error is indeed reported asynchronously in json-rpc mode. I do need to think about it a bit how to actually implement that. In general it would be fine for me to report that error back via a webhook. The only problem I see here is, that a webhook is most probably too complicated for some users (there is a significant portion of users who use the API in Home Assistant and do not have that much technical knowledge and they would struggle to grasp the concept of webhooks). I am bit worried that this would then result in quite a lot of support issues :/ 

Maybe a error endpoint that could be polled would be an alternative? Or a error reporting via websocket? I don't know...I guess I'll need to think about it a bit.

btw: I don't know if you actually need the websocket receive, but in case you do not need it, one alternative would be to switch to the native mode. That one should work reasonable fast and you should get back the challenge token immediately in case the message can't be sent. Of course this is just a workaround until the functionality is implemented in the json-rpc mode.

**Comment by benjy3gg** [comment]
getting the data back immediatly would be best, but if the error is returned asynchronously in the api its a bit harder to implement on my end. The webhook would be the nicest way to fix the rate limit challenge from our usage.
An endpoint that returns the rate limit errors (and all other async errors) would also be fine. 
For that to be usable, you would probably need to add the request id to the 400 error message:
`{ "error": "Failed to send message due to rate limiting", "id": "id-here-uuid-thing"}`
to be able to match it later on with the data in the endpoint.

**Comment by bbernhard** [comment]
Yeah, it is a pity, that signal-cli doesn't return the full error synchronously. That would make it much easier.

How time critical is your use case? Let's say we go for the most simple solution and expose an endpoint for that (e.g `/v1/<number>/errors`). Would it be fine for you to poll that endpoint or is that too slow?

I can imagine the following API:

`GET /v1/<number>/errors` returns a list with all the errors, each error has it's own identifying error uuid

Either we automatically clear the errors once someone has fetched them via the GET endpoint, or we could introduce another API endpoint to clear individual error messages. So something like:

`DELETE /v1/<number>/errors/<error uuid>`

So, once you've handled/acknowledged the error, you could remove it, so that it isn't returned again with the next GET.

Do you need persistence? Is it fine for your use case, if the errors are kept in memory (and would be gone when the docker container restarts) or should the errors be persistent (i.e written to the filesystem)?

I am just trying to understand your use case better and gathering some real world input :) 

> For that to be usable, you would probably need to add the request id to the 400 error message

jep, right.

**Comment by benjy3gg** [comment]
we have a cronjob that runs every 20mins anyway to retry sending the message. 
So an endpoint that returns the errors with tokens would be perfect, as we could just try to query the endpoint and check if there as an associated error.
Peristence would be a plus but i think its not critical.

regarding the deletion method:
do you know how long the challenge tokens are valid and also, if the challenge token is changing when you try to send a message to the same recipient again? (hope its clear what i mean).
If the challenge token stays the same, the DELETE would be better.
otherwise deleting the error on fetch would be nicer.

**Comment by bbernhard** [comment]
> we have a cronjob that runs every 20mins anyway to retry sending the message. So an endpoint that returns the errors with tokens would be perfect, as we could just try to query the endpoint and check if there as an associated error. Peristence would be a plus but i think its not critical.
> 
> regarding the deletion method: do you know how long the challenge tokens are valid and also, if the challenge token is changing when you try to send a message to the same recipient again? (hope its clear what i mean). If the challenge token stays the same, the DELETE would be better. otherwise deleting the error on fetch would be nicer.

Got it, thanks!

I think those challenge tokens are only short-lived, but unfortunately I don't know how long they are exactly valid.

But _maybe_ I've found another, better solution, which makes it possible to already return the challenge token as a response to the failing send request. Unfortunately I am not able to really test that (as I can't provoke the situation myself).

So maybe you could give `bbernhard/signal-cli-rest-api:0.150-dev` a try

**Comment by benjy3gg** [comment]
`
<img width="779" alt="image" src="https://github.com/bbernhard/signal-cli-rest-api/assets/458675/26f3acf8-62f8-4727-86ab-f9f7c609325d">
`
looks good, could you put the token under its own key in the json response?

**Comment by bbernhard** [comment]
yeah, sure. Just wanted to create a version with some minimal changes first to see if it works like intended :)

**Comment by benjy3gg** [comment]
perfect!

**Comment by bbernhard** [comment]
Please try `bbernhard/signal-cli-rest-api:0.151-dev`

**Comment by benjy3gg** [comment]
with 151-dev i get the old error message back:
<img width="413" alt="image" src="https://github.com/bbernhard/signal-cli-rest-api/assets/458675/f72e28b4-56db-4049-a3b9-1c8d4416451a">

**Comment by bbernhard** [comment]
Oh, damn. Can you please try `bbernhard/signal-cli-rest-api:0.152-dev`

**Comment by benjy3gg** [comment]
<img width="790" alt="image" src="https://github.com/bbernhard/signal-cli-rest-api/assets/458675/1eda2b42-ccbd-42b8-8343-3a31af38f7af">

**Comment by bbernhard** [comment]
Please try with `bbernhard/signal-cli-rest-api:0.153-dev`. In case this still doesn't work, then please enable debug logging again. I've added some debug log statements this time.

**Comment by benjy3gg** [comment]
<img width="777" alt="image" src="https://github.com/bbernhard/signal-cli-rest-api/assets/458675/39008204-3410-4473-9bc5-505bbf3bc768">

**Comment by bbernhard** [comment]
> <img alt="image" width="777" src="https://private-user-images.githubusercontent.com/458675/317781964-39008204-3410-4473-9bc5-505bbf3bc768.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE2NTA4MzgsIm5iZiI6MTcxMTY1MDUzOCwicGF0aCI6Ii80NTg2NzUvMzE3NzgxOTY0LTM5MDA4MjA0LTM0MTAtNDQ3My05YmM1LTUwNWJiZjNiYzc2OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyOFQxODI4NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZWZiNTZjNzQwN2E5ZmUzNDg0YTkzMTAzZGY4YmM1YWU2NTUyYThjZTU1ODRkNDFiZGE2OTA0NGI5MmYzMzM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.l6xsSo6mWnNBLZZQ47vkRgXJBIGgnRtwWudbABak1q8"> looks good now.
> 
> only thing that would be nice now, would be if the json also contained a "status" field with the correct http status code :D

Great that it is working now!

Could you elaborate a bit more what you exactly mean? Right now the REST request should fail with HTTP code 400. I've just changed it to return 429 (Too Many Requests) in that case. Is that what you meant?

**Comment by benjy3gg** [comment]
yes thats what i want. when will you merge this into release?

**Comment by bbernhard** [comment]
should already be in the latest release

**Comment by martonejd** [comment]
Hi,
I'm getting this error. Can you give me an idea of what I need to do with the challenge token? Do I insert it in the original curl command? If so, can you provide an example of the syntax? I'm not a developer, if it's more complicated than that, but a general idea of what to do would help. I can try to search for answers after that.
Thanks!

**Comment by bbernhard** [comment]
see https://github.com/bbernhard/signal-cli-rest-api/issues/488#issuecomment-1960075430
